File: solutions/0581. Shortest Unsorted Continuous Subarray/0581.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
 public:
  int findUnsortedSubarray(const vector<int>& nums) {
    const int n = nums.size();
    int mn = INT_MAX;
    int mx = INT_MIN;
    bool meetDecrease = false;
    bool meetIncrease = false;

    for (int i = 1; i < n; ++i) {
      if (nums[i] < nums[i - 1])
        meetDecrease = true;
      if (meetDecrease)
        mn = min(mn, nums[i]);
    }

    for (int i = n - 2; i >= 0; --i) {
      if (nums[i] > nums[i + 1])
        meetIncrease = true;
      if (meetIncrease)
        mx = max(mx, nums[i]);
    }

    int l = 0;
    while (l < n && nums[l] <= mn) {
      ++l;
    }

    int r = n - 1;
    while (r >= 0 && nums[r] >= mx) {
      --r;
    }

    return l < r ? r - l + 1 : 0;
  }
};

File: solutions/0581. Shortest Unsorted Continuous Subarray/0581.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        maxi = nums[0]
        last = -1
        
        for i in range(len(nums)):
            maxi = max(maxi, nums[i])
            if nums[i] < maxi:
                last = i
        
        if last == -1:
            return 0
        
        mini = nums[-1]
        first = -1
        for i in range(len(nums)-1, -1, -1):
            mini = min(mini, nums[i])
            if nums[i] > mini:
                first = i
        
        return last - first + 1